### PR TITLE
RFC: SoundFont Loading Refactor

### PR DESCRIPTION
**Pull request by mawe42**
Important: This isn't a finished pull request, but a work in progress. So please do not merge this in it's current form.

I've started this large refactor because I wanted to add a new feature to the soundfont loader: lazy-loading of presets and samples. The idea is that in certain real-time performance situations, small delays in switching presets for loading the required samples are absolutely acceptable. So on loading a new soundfont and selecting a preset, only the absolute minimum of information is actually read from the file and loaded to RAM. This would greatly reduce loading time and save huge amounts of memory for larger soundfonts that contain many instruments and presets. Please note: this feature hasn't been implemented in this PR. I'm just explaining this here to give some background to some of the changes I've made to the loading infrastructure.

Just a quick summary of what has already been done:
- more the low-level soundfont loading code (the code taken over from Smurf) into a separate file
- cleanup the low-level code, bringing the style more in line with the rest of FluidSynth's codebase
- refactor the low-level code to prepare it for selective loading of preset zones, instruments and samples
- move the ogg vorbis compression to fluid_sfont
- refactor the sample start/end and loopstart/loopend checking, making it possible to use the same logic
and code for both compressed and uncompressed samples
- move the samplecache code from defsfont to own file, move sample loading code to low-level code in fluid_sffile
- refactor and cleanup the samplecache code

Next I am going to work in the preset loading functions. But before I carry on and add more hours and work, I would really like to get some feedback on the direction I'm taking. Do the changes I've made so far look good, or do you think I'm going into the wrong direction?

Also, I've got one specific question: currently, whenever FluidSynth core requests a preset from a defsfont  loader, a new fluid_preset_t struct is allocated... every time. To me that seems totally pointless, because that fluid_preset_t struct only contains pointers to the accessor functions and uses the (singleton) fluid_defpreset_t to do any real work anyway. So I'm thinking of changing the defsfont loader to pre-create those fluid_preset_t structs and only return a pointer to them when asked. That would mean we could get rid of the preset_stack dance in defspreset.

### CODE REVIEW

File: src/sfloader/fluid_sfont.c [hunk 1]
```diff
@@ -546,3 +551,242 @@ int fluid_sample_set_pitch(fluid_sample_t* sample, int root_key, int fine_tune)
...
+     * correct. Hours of thinking through this have concluded that it would be
+     * best practice to mangle with loops as little as necessary by only making
+     * sure the pointers are within sample->start to max_end. Incorrect
+     * soundfont shall preferably fail loudly. */
+    if ((sample->loopstart < sample->start) || (sample->loopstart > max_end))
+    {
+        FLUID_LOG(FLUID_DBG, "Sample '%s': invalid loop start '%d', setting to sample start '%d'",
+                sample->name, sample->loopstart, sample->start);
+        sample->loopstart = sample->start;
+        modified = TRUE;
```
Comment (derselbst) [inline src/sfloader/fluid_sfont.c]: This actually introduces a regression for #171. As the comment points out we should not modify this in case of `(sample->loopstart < sample->start)`, at least for SF2 files. This handling would be probably ok for SF3, as it does not share a big sample buffer like SF2.

File: src/sfloader/fluid_sfont.h [hunk 2]
```diff
@@ -24,6 +24,10 @@
 
 #include "fluidsynth.h"
 
+FLUIDSYNTH_API int fluid_sample_validate(fluid_sample_t *sample, unsigned int max_end);
+FLUIDSYNTH_API int fluid_sample_sanitize_loop(fluid_sample_t *sample, unsigned int max_end);
+FLUIDSYNTH_API int fluid_sample_decompress_vorbis(fluid_sample_t *sample);
```
Comment (derselbst) [inline src/sfloader/fluid_sfont.h]: Why are you making these public?

File: src/sfloader/fluid_sfont.c (hunk 1 above)
Comment (mawe42) [reply src/sfloader/fluid_sfont.c]: I don't think it introduces a regression. It's not quite what the comment says, but that's how it is implemented in master:
```
invalid_loopstart = (sam->loopstart < sam->start) || (sam->loopstart >= sam->loopend);
```

And when invalid_loopstart is set, loopstart is set to sample->start.

File: src/sfloader/fluid_sfont.h (hunk 2 above)
Comment (mawe42) [reply src/sfloader/fluid_sfont.h]: Ah, good point. That's a mistake.

File: src/sfloader/fluid_sfont.c (hunk 1 above)
Comment (derselbst) [reply src/sfloader/fluid_sfont.c]: Sry, my bad. I was convinced it sounded wrong during a first rehearsal. But you're right, it's fine.

File: src/sfloader/fluid_sffile.h [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+                 equivalent to loopstart */
+    unsigned int samplerate; /* Sample rate recorded at */
+    unsigned char origpitch; /* root midi key number */
+    signed char pitchadj; /* pitch correction in cents */
+    unsigned short sampletype; /* 1 mono,2 right,4 left,linked 8,0x8000=ROM */
+};
+
+struct _SFInst
+{ /* Instrument structure */
+    unsigned short idx;
```
Comment (derselbst) [inline src/sfloader/fluid_sffile.h]: Perhaps better use `unsigned int` for this index?

File: src/sfloader/fluid_sfont.h [hunk 4]
```diff
@@ -170,6 +174,7 @@ struct _fluid_preset_t {
  */
 struct _fluid_sample_t
 {
+  unsigned short idx;
```
Comment (derselbst) [inline src/sfloader/fluid_sfont.h]: `uint idx;` ?

File: src/sfloader/fluid_samplecache.c [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    }
+
+    return NULL;
+}
+
+static int fluid_get_file_modification_time(char *filename, time_t *modification_time)
+{
+#if defined(WIN32) || defined(__OS2__)
+    *modification_time = 0;
+    return FLUID_OK;
```
[the remark this answers is not in the document]
Comment (derselbst) [reply src/sfloader/fluid_samplecache.c]: Here's a quickly smashed together C99 version that should retrieve the mod time of a file. Uncompiled and untested, as I'm currently not at home.

```c
int res = FLUID_FAILED;

#if GLIB_CHECK_VERSION(2,26,0) // g_date_time_new_from_timeval_utc() requires at least 2.26
GFile* gfile = g_file_new_for_path (filename);

if(gfile != NULL)
{
GFileInfo* info = g_file_query_info (gfile,
                                  G_FILE_ATTRIBUTE_STANDARD_SIZE "," G_FILE_ATTRIBUTE_TIME_MODIFIED,
                                  G_FILE_QUERY_INFO_NONE,
                                  NULL,
                                  NULL);
if (info != NULL)
{
                GTimeVal tv;
                g_file_info_get_modification_time (info, &tv);
                GDateTime* dt = g_date_time_new_from_timeval_utc(&tv);
                if(dt != NULL)
                {
                   // TODO: converting int64 to time_t, does it compile? is it safe??
                   *modification_time = g_date_time_to_unix(dt);
                    g_date_time_unref(dt);
                    res = FLUID_OK;
                }
                g_object_unref(info);
}
g_object_unref(gfile);
}
#else
#warning "Consider upgrading glib to at least version 2.26"
#endif
return res;
```

File: src/sfloader/fluid_samplecache.c [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    }
+
+    if (try_mlock && !entry->mlocked)
+    {
+        /* Lock the memory to disable paging. It's okay if this fails. It
+         * probably means that the user doesn't have the required permission. */
+        entry->mlocked = (fluid_mlock(entry->sample_data, entry->sample_count * 2) == 0);
+        if (entry->sample_data24 != NULL)
+        {
+            entry->mlocked &= (fluid_mlock(entry->sample_data24, entry->sample_count) == 0);
```
[the remark this answers is not in the document]
Comment (mawe42) [reply src/sfloader/fluid_samplecache.c]: Good point! Will change it so that it only tries to the mlock 24-bit data if the first mlock succeeds.

File: src/sfloader/fluid_samplecache.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mawe42) [reply src/sfloader/fluid_samplecache.c]: Great, thanks! Regarding the g_date_time_new_from_timeval_utc: we could switch to using a uint64 for the modification_time in fluid_samplecache_t. Then we could just to the following:
```
*modification_time = tv.tv_sec * 1000000 + tv.tv_usec;
```
What do you think?

File: src/sfloader/fluid_sffile.c [hunk 7]
```diff
@@ -0,0 +1,1797 @@
...
+    if (sf->sample24pos && ((start > sf->sample24size) || (end > sf->sample24size)))
+    {
+        FLUID_LOG(FLUID_ERR, "Sample offsets exceed 24-bit sample data chunk");
+        goto error_exit;
+    }
+
+    /* Load 16-bit sample data */
+    if (sf->fcbs->fseek(sf->sffd, sf->samplepos + (start * 2), SEEK_SET) == FLUID_FAILED)
+    {
+        perror("error");
```
[the remark this answers is not in the document]
Comment (mawe42) [reply src/sfloader/fluid_sffile.c]: Ack

File: src/sfloader/fluid_sffile.c [hunk 8]
```diff
@@ -0,0 +1,1797 @@
...
+        FLUID_LOG(FLUID_ERR, "Failed to seek position in data file");
+        goto error_exit;
+    }
+
+    loaded_data = (short *)FLUID_MALLOC(count * 2);
```
[the remark this answers is not in the document]
Comment (mawe42) [reply src/sfloader/fluid_sffile.c]: Ok, I'll use FLUID_ARRAY.

File: src/sfloader/fluid_sffile.c [hunk 9]
```diff
@@ -0,0 +1,1797 @@
...
+
+    end = start + count;
+
+    if (((start * 2) > sf->samplesize) || ((end * 2) > sf->samplesize))
+    {
+        FLUID_LOG(FLUID_ERR, "Sample offsets exceed sample data chunk");
+        goto error_exit;
+    }
+
+    if (sf->sample24pos && ((start > sf->sample24size) || (end > sf->sample24size)))
```
[the remark this answers is not in the document]
Comment (mawe42) [reply src/sfloader/fluid_sffile.c]: Yes, good idea. And more in line with how we treat 24-bit samples everywhere else.

File: src/sfloader/fluid_sffile.c [hunk 10]
```diff
@@ -0,0 +1,1797 @@
...
+            perror("error");
+            FLUID_LOG(FLUID_ERR, "Failed to seek position in data file");
+            goto error_exit;
+        }
+
+        loaded_data24 = (char *)FLUID_MALLOC(count);
+        if (loaded_data24 == NULL)
+        {
+            FLUID_LOG(FLUID_ERR, "Out of memory");
+            return FLUID_FAILED;
```
[the remark this answers is not in the document]
Comment (mawe42) [reply src/sfloader/fluid_sffile.c]: Ack

File: src/sfloader/fluid_sffile.h (hunk 3 above)
Comment (mawe42) [reply src/sfloader/fluid_sffile.h]: Yes... no idea why I chose short here...

File: src/sfloader/fluid_sfont.h (hunk 4 above)
Comment (mawe42) [reply src/sfloader/fluid_sfont.h]: Yes!

File: src/sfloader/fluid_samplecache.c [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+    fluid_mutex_lock(samplecache_mutex);
+
+    entry = get_samplecache_entry(sf, sample_start, sample_count);
+    if (entry == NULL)
+    {
+        entry = new_samplecache_entry(sf, sample_start, sample_count);
+        if (entry == NULL)
+        {
+            return FLUID_FAILED;
```
[the remark this answers is not in the document]
Comment (mawe42) [reply src/sfloader/fluid_samplecache.c]: Well spotted, thanks.

File: src/sfloader/fluid_sffile.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (mawe42) [reply src/sfloader/fluid_sffile.c]: Hm... seeing that FLUID_ARRAY does exactly that (just hidden behind a macro)... what exactly do you propose?

[the remark this answers is not in the document]
Comment (mawe42) [reply src/sfloader/fluid_sffile.c]: Ah, now I understand. You just want to remove the explicit cast?

**Comment by mawe42** [comment]
> Overall this looks good to me. It splits the huge overloaded fluid_defsfont.c into more handy parts.

Great, thanks for your feedback and the initial review! I'll carry on then :-) There are a few more refactorings I'd like to do to the remaining defsfont code before attempting to add the new feature.

> It might be worth bringing the lazy-loading feature to the mailing list for discussion, but that's a different story.

Ok, I will do that.

> Not quite sure about the schedule of your works. Would it be #360 -> Re-Formatting -> Preset-Stack-Refactor -> Lazy-Loading ? I'm open to postpone the Reformatting even further, just let me know.

It would be great if we could postpone the reformatting until after the additional refactorings I'd like to do. So #360 -> Preset-Stack + a few other refactorings -> Re-Formatting -> Lazy Loading.

File: src/sfloader/fluid_samplecache.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (derselbst) [reply src/sfloader/fluid_samplecache.c]: Ok sure. But unless you really need microsecond precision for a file mod date, `*modification_time = tv.tv_sec` should be sufficient (which should result in quite the same behaviour as g_date_time_to_unix()).

[the remark this answers is not in the document]
Comment (mawe42) [reply src/sfloader/fluid_samplecache.c]: Hm, microsecond precision is probably not required. So lets go with seconds and see how it goes. Precision can always be easily increased if it poses a problem...

File: src/sfloader/fluid_sffile.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (derselbst) [reply src/sfloader/fluid_sffile.c]: > You just want to remove the explicit cast?

Yes. But FLUID_ARRAY really seems to be the better solution for this purpose (just forgot about it).

> seeing that FLUID_ARRAY does exactly that (just hidden behind a macro)...

At least the cast is "abstracted" through the macro, so I dont further care.

**Comment by mawe42** [comment]
I like your idea of only lazy loading sample data first. So I'm going to rebase this branch again, trying to remove most of the functional changes I made to the low-level loading in fluid_sffile that were in preparation for the lazy-loading of structural information. Should have kept this as a clean refactor in the first place and not mix it with functional changes... :-/

Then I will add the new modification time code and get rid of the preset stack and close this branch. Lazy loading will come in a new branch (and possibly after reformatting). Does that sound good to you, or would you rather have the preset stack stuff in a separate PR?

File: src/sfloader/fluid_samplecache.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (derselbst) [reply src/sfloader/fluid_samplecache.c]: Oh, I didnt notice I was digging in GIO. You're right, and `g_stat` is the much more elegant way. Thx!

**Comment by derselbst** [comment]
> So I'm going to rebase this branch again, trying to remove most of the functional changes I made to the low-level loading in fluid_sffile

Ok. But generally I like the idea of having this fluid_sffile to handle the low-level loading.

If it's not too much trouble for you I would choose to have the preset stack separately. Apart from that this seems fine to me.

**Comment by derselbst** [comment]
Ok will have a look at it. And just forget about the reformat at least for now, currently too much activity. Will see when it fits.

**Comment by derselbst** [comment]
No, fluidsynth requires 2.6.5 anyway:

https://github.com/FluidSynth/fluidsynth/blob/91966361770930ae0873b0fb38241eabcb89b655/CMakeLists.txt#L384

**Comment by mawe42** [comment]
> No, fluidsynth requires 2.6.5 anyway

Perfect.

File: src/sfloader/fluid_samplecache.c [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+                if (entry->mlocked)
+                {
+                    fluid_munlock(entry->sample_data, entry->sample_count * 2);
+                    if (entry->sample_data24 != NULL)
+                    {
+                        fluid_munlock(entry->sample_data24, entry->sample_count);
+                    }
+                }
+
+                fluid_list_remove(samplecache_list, entry);
```
Comment (derselbst) [inline src/sfloader/fluid_samplecache.c]: This should be `samplecache_list = fluid_list_remove(samplecache_list, entry);`. Otherwise `samplecache_list` is not set to NULL when the last element is removed resulting in a NULL deref in line 248.

File: src/sfloader/fluid_sfont.c [hunk 13]
```diff
@@ -546,3 +551,242 @@ int fluid_sample_set_pitch(fluid_sample_t* sample, int root_key, int fine_tune)
...
+        buffer_size /= 2;
+    }
+
+    if ((sample->end > buffer_size) || (sample->start >= sample->end))
+    {
+        FLUID_LOG(FLUID_WARN, "Sample '%s': invalid start/end file positions", sample->name);
+        return FLUID_FAILED;
+    }
+
+    return TRUE;
```
Comment (derselbst) [inline src/sfloader/fluid_sfont.c]: Because you're using FLUID_FAILED, this should be FLUID_OK.

File: src/sfloader/fluid_samplecache.c (hunk 12 above)
Comment (mawe42) [reply src/sfloader/fluid_samplecache.c]: Of course, well spotted.

File: src/sfloader/fluid_sfont.c (hunk 13 above)
Comment (mawe42) [reply src/sfloader/fluid_sfont.c]: Good point.

**Comment by derselbst** [comment]
So far testing worked as expected, no problems nor crashes. I also think it's time to add some unittest(s). I added one for the sample cache (#361), mainly aiming to test the ref count mechanism, i.e. to avoid using sample data from a soundfont that has already been unloaded by its parent synth.

There is a `struct _fluid_sfont_info_t` in fluid_synth.h which seems to be a bit redundant as it adds another indirection layer. I think it's members could just be integrated in `struct _fluid_sfont_t` as it's not a public struct anymore. And that `fluid_synth_t::sfont_hash` seems to be completely redundant? (just by passing the `sfont_info` directly
 
```diff
- fluid_synth_sfont_unref (synth, sfont_info->sfont);
+ fluid_synth_sfont_unref (synth, sfont_info);
```

in fluid_synth.c???)

Long story short: This PR is fine. If you want you can take care of those two minor issues right here or I'll have a try.

**Comment by derselbst** [comment]
> Maybe he didn't want to change fluid_sfont_t because it was public. And the hashtable to avoid having to iterate a list when searching for the fluid_sfont_info_t for a sfont?

Just my thought.

**Comment by derselbst** [comment]
Just found that the unit test segfaulted with the vintagedreams sfont using your changes. Fixed a misplaced `return FLUID_OK` and modified the sample import code to not exit early, when it sees a broken sample:

32961c4031249f1232c17b3e4bc6e4781f87df38...6ab1c7fde32e172f81103a86b7301af4b3029cc7

**Comment by mawe42** [comment]
Ah, thanks for catching that!

**Comment by mawe42** [comment]
I've just removed the noise debug messages for samples with loopstart = loopend: fa0d10390705cd6eba0e2d5c8db28d0bd66bd715
